src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -19,6 +19,8 @@ public partial class Ping : Component
         private SendOrPostCallback? _onPingCompletedDelegate;
         private bool _disposeRequested;
         private byte[]? _defaultSendBuffer;
+        private CancellationTokenSource? _timeoutOrCancellationSource;
+        // Used to differentiate between timeout and cancellation when _timeoutOrCancellationSource triggers
         private bool _canceled;
```

Comment:
I'm surprised that this doesn't need to be `volatile` since it gets written to and read from on different threads, with the reads eschewing a lock. With the updated code the read pattern is unchanged (no lock), we now write to this outside of a lock sometimes (from the cancellation token registration); does that change anything?

Reply:
Since this is always set together with cancelling the _`timeoutOrCancellationSource`, I think we can get away without volatile, but I will defer to the opinion of the more experienced. cc: @stephentoub

Reply:
It doesn't. There's always a happens-before relationship involved, e.g. the flag is set before the asynchronous operation is started, or it's read only after the things that could have set it have demonstrably quiesced.